File: sourcecodes/binarysearch/binarysearch.py
import random
import typing

class ParameterError(Exception):
    pass

def binarysearch(array : list, target): #target int or float
    if not array:
        raise ParameterError("List cannot be empty")
    elif not (type(target) == int or type(target) == float):
        raise ParameterError("Target has to be either of type int or float")
    elif array[0] == target:
        rv.index = 0
        rv.l = 0
        for i in array:
            if i != target: break
            else: rv.l += 1
        
    rv = search(array, target) #return value
    array.sort()
    h = h2 = HI = len(array)-1 #the highest index of the array 
    l = mid = l2 = mid2 =0 #l = lowest (index)
    while l+1 != h:
        mid = (h+l)//2
        mid2 = (h2+l2)//2
        if array[mid] < target:
            l = mid
        else:
            h = mid
        if array[mid2] <= target:
            l2 = mid2
        else:
            h2 = mid2
            
    if array[h] != target:
        rv.index = False 
    else:
        rv.index = h #index of first instance of target
    rv.l = (l2-h)+1 #calculate instances of same number
    rv.greater = HI-l2 #amount of numbers greater than target
    
    rv.lesser = h  #amount of numbers lesser than target
    
            

class search:
    def __init__(self, array, target):
        self.array = array
        self.target = target
        self.index = 0 #the index of first instance of target
        self.l = 0 #amount of targets in array
    
    def __len__(self):
        return self.l
    
     

File: sourcecodes/binarysearch/cpp/binarysearch.cc
#include <iostream>
#include <string>
#include <sstream>
#include <algorithm>
using namespace std;

class ParameterError {
};

class searched {
    public:
        int greater;
        int smaller;
        int index;
        int num;
        
};

searched binarysearch(int array[], int target) 
{
    int n = sizeof(array) / sizeof(array[0]);
    cout << "the size of n is " << n << "\n";
    if (n == 0) {
        throw ParameterError();
    } 
    searched rv;  // return value
    sort(array, array+n);
    for (int i = 0;i<n;i++) {
        cout << array[i] << "\n";
    }
    int l = 0;
    int mid = 0;
    int l2 = 0;
    int mid2 = 0;
    int h = n-1; 
    int h2 = n-1;
    int HI = n-1;
    while (l+1 != h) {
        mid = (l+h)/2;
        mid2 = (l2+h2)/2;
        if (array[mid] < target) {
            l = mid;
        } else {
            h = mid;
        }
        if (array[mid2] <= target) {
            l2 = mid2;
        } else {
            h2 = mid2;
        }
    }
    if (array[h] != target) {
        rv.index = false;
    } else {
        rv.index = h;   
    }
    rv.num = (l2-h)+1; //calculate instances of same number
    rv.greater = HI-l2; //amount of numbers greater than target
    
    rv.smaller = h;
    return rv;
}
int main()
{
    ios_base::sync_with_stdio(false);
    int test[7] = {1, 5, 2, 3, 7, 7, 2};
    searched rv = binarysearch(test, 5);
    cout << "ended" << "\n";
    cout << rv.index << "\n";
    cout << "outted index" << "\n";
    cout << rv.num;
}
